File: server/src/services/normalizer.ts
// =============================================================================
// Normalizer — converts raw YouTube API responses to CanonicalSocialEvent
// =============================================================================

import type { CanonicalSocialEvent } from '../types/index.js';
import type { youtube_v3 } from 'googleapis';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Extract #hashtags from a string. */
function extractHashtags(text: string): string[] {
  const matches = text.match(/#[\w\u0900-\u097F]+/g); // supports Latin + Devanagari
  return matches ? [...new Set(matches.map((h) => h.toLowerCase()))] : [];
}

/** Extract @mentions from a string. */
function extractMentions(text: string): string[] {
  const matches = text.match(/@[\w.]+/g);
  return matches ? [...new Set(matches.map((m) => m.toLowerCase()))] : [];
}

/** Generate a deterministic event_id for YouTube content. */
function makeEventId(prefix: string, id: string): string {
  return `yt_${prefix}_${id}`;
}

// ---------------------------------------------------------------------------
// Video → CanonicalSocialEvent
// ---------------------------------------------------------------------------

/**
 * Maps a YouTube video (search/video item + statistics) to a CanonicalSocialEvent.
 */
export function videoToCanonicalEvent(
  video: youtube_v3.Schema$Video,
): CanonicalSocialEvent {
  const snippet = video.snippet!;
  const stats = video.statistics;
  const videoId = video.id!;

  const title = snippet.title ?? '';
  const description = snippet.description ?? '';
  const fullText = `${title}\n\n${description}`;

  // Combine tags from snippet + hashtags extracted from description
  const snippetTags = (snippet.tags ?? []).map((t) => `#${t.toLowerCase()}`);
  const descriptionTags = extractHashtags(description);
  const allHashtags = [...new Set([...snippetTags, ...descriptionTags])];

  return {
    event_id: makeEventId('video', videoId),
    platform: 'youtube',
    event_type: 'post',
    source: {
      source_id: videoId,
      url: `https://www.youtube.com/watch?v=${videoId}`,
      collector: 'youtube_data_api_v3',
    },
    author: {
      user_id: snippet.channelId ?? 'unknown',
      username: snippet.channelId ?? 'unknown',
      display_name: snippet.channelTitle ?? 'Unknown Channel',
    },
    content: {
      text: fullText,
      language: snippet.defaultLanguage ?? snippet.defaultAudioLanguage ?? 'en',
      hashtags: allHashtags,
      mentions: extractMentions(fullText),
    },
    engagement: {
      likes: parseInt(stats?.likeCount ?? '0', 10),
      comments: parseInt(stats?.commentCount ?? '0', 10),
      shares: 0, // YouTube API does not expose share count
      views: parseInt(stats?.viewCount ?? '0', 10),
    },
    relationships: {
      reply_to: null,
      repost_of: null,
      quoted_event_id: null,
    },
    timestamps: {
      created_at: snippet.publishedAt ?? new Date().toISOString(),
      collected_at: new Date().toISOString(),
    },
    analysis: {
      sentiment: { label: 'neutral', score: 0 },
      emotion: { label: 'surprise', score: 0 },
      stance: { label: 'neutral', score: 0 },
      sarcasm: { detected: false, score: 0 },
    },
    collection_reason: ['youtube_search_ingestion'],
  };
}

// ---------------------------------------------------------------------------
// Comment → CanonicalSocialEvent
// ---------------------------------------------------------------------------

/**
 * Maps a YouTube comment (top-level or reply) to a CanonicalSocialEvent.
 */
export function commentToCanonicalEvent(
  comment: youtube_v3.Schema$Comment,
  videoId: string,
  isReply: boolean = false,
  parentCommentId?: string,
): CanonicalSocialEvent {
  const snippet = comment.snippet!;
  const commentId = comment.id!;
  const text = snippet.textOriginal ?? snippet.textDisplay ?? '';

  return {
    event_id: makeEventId('comment', commentId),
    platform: 'youtube',
    event_type: isReply ? 'reply' : 'comment',
    source: {
      source_id: commentId,
      url: `https://www.youtube.com/watch?v=${videoId}&lc=${commentId}`,
      collector: 'youtube_data_api_v3',
    },
    author: {
      user_id: snippet.authorChannelId?.value ?? 'unknown',
      username: snippet.authorChannelId?.value ?? 'unknown',
      display_name: snippet.authorDisplayName ?? 'Anonymous',
      avatarUrl: snippet.authorProfileImageUrl ?? undefined,
    },
    content: {
      text,
      language: 'en', // YouTube comment API does not return language
      hashtags: extractHashtags(text),
      mentions: extractMentions(text),
    },
    engagement: {
      likes: snippet.likeCount ?? 0,
      comments: 0,
      shares: 0,
    },
    relationships: {
      reply_to: isReply ? (parentCommentId ?? videoId) : videoId,
      repost_of: null,
      quoted_event_id: null,
    },
    timestamps: {
      created_at: snippet.publishedAt ?? new Date().toISOString(),
      collected_at: new Date().toISOString(),
    },
    analysis: {
      sentiment: { label: 'neutral', score: 0 },
      emotion: { label: 'surprise', score: 0 },
      stance: { label: 'neutral', score: 0 },
      sarcasm: { detected: false, score: 0 },
    },
    collection_reason: ['youtube_comment_ingestion'],
  };
}

File: backend_python/app/x/normalizer.py
"""
X Normalizer module.
Maps parsed raw tweet dicts into CanonicalSocialEvent models.
"""

from datetime import datetime
from typing import Dict, Any
from app.models.social_event import (
    CanonicalSocialEvent,
    SocialAuthor,
    EventEngagement,
    EventSource,
    EventContent,
    EventRelationships,
    EventTimestamps,
    EventAnalysis,
)


def normalize_x_tweet(raw_tweet: Dict[str, Any], query: str = "") -> CanonicalSocialEvent:
    """Map parsed X tweet dictionary into CanonicalSocialEvent schema."""
    tweet_id = str(raw_tweet.get("tweet_id", ""))
    is_reply = raw_tweet.get("is_reply", False)
    parent_id = raw_tweet.get("parent_tweet_id")

    event_id = f"x_reply_{tweet_id}" if is_reply else f"x_post_{tweet_id}"
    event_type = "reply" if is_reply else "post"

    created_at = raw_tweet.get("created_at") or datetime.utcnow().isoformat() + "Z"

    return CanonicalSocialEvent(
        event_id=event_id,
        platform="x",
        event_type=event_type,
        source=EventSource(
            source_id=tweet_id,
            url=raw_tweet.get("tweet_url", f"https://x.com/status/{tweet_id}"),
            collector="x_playwright",
        ),
        author=SocialAuthor(
            user_id=raw_tweet.get("author_id", "unknown"),
            username=raw_tweet.get("username", "unknown"),
            display_name=raw_tweet.get("display_name", "X User"),
            avatarUrl=raw_tweet.get("avatar_url"),
            followers_count=raw_tweet.get("followers_count"),
        ),
        content=EventContent(
            text=raw_tweet.get("text", ""),
            language="en",
            hashtags=raw_tweet.get("hashtags", []),
            mentions=raw_tweet.get("mentions", []),
        ),
        engagement=EventEngagement(
            likes=raw_tweet.get("likes") or 0,
            comments=raw_tweet.get("comments") or 0,
            shares=raw_tweet.get("shares") or 0,
            views=raw_tweet.get("views"),
        ),
        relationships=EventRelationships(
            reply_to=parent_id if is_reply else None,
            parent_post_id=parent_id if is_reply else None,
            mentions=raw_tweet.get("mentions", []),
        ),
        timestamps=EventTimestamps(
            created_at=created_at,
            collected_at=datetime.utcnow().isoformat() + "Z",
        ),
        analysis=EventAnalysis(),
        collection_reason=["recent", "relevant"] if query else ["recent"],
    )
